src/rust/engine/process_execution/src/remote.rs

```diff
@@ -88,7 +82,7 @@ impl CommandRunner {
   fn oneshot_execute(
     &self,
     execute_request: bazel_protos::build::bazel::remote::execution::v2::ExecuteRequest,
-  ) -> impl Future<Item = OperationOrStatus, Error = String> {
+  ) -> impl Future<Item = bazel_protos::google::longrunning::Operation, Error = String> {
```

Comment:
Out of curiosity, what is the utility of using `impl Future<Item = .., Error = ...>` here? Is the idea that the method can return anything type convertible via `From`/`Into` into these types (or does it allow us to avoid calling `.to_boxed()` on the result?)? Mostly asking to see if I should be writing this on all my futures-y code.

Reply:
`impl Future` is shorthand for "Exactly one type which implements the `Future` trait".

So we can't use it in places like:
```rust
fn... {
  if x > 1 {
    futures::future::ok("Hello")
  } else {
    futures::future::ok("goodbye").map(str::to_upper)
  }
}
```

because the function returns two different types that can't be unified.

`BoxFuture` is used to make both of the things we return have the same type, so that we can have a single return type. It does this by moving the `dyn Future` to the heap, and making a `BoxFuture` pointing to it. We introduced `BoxFuture` before `impl Future` was a stable thing in the language; we should prefer `impl Future` to `BoxFuture` because it avoids moving the `Future` to the heap. But `BoxFuture` still currently has its place; mostly it's used where it's much more clear to early return than the use awkward combinators:

```rust
fn... {
  if some_condition {
    return futures::future::err("Broken").to_boxed();
  }
  some_complex_future_expression().to_boxed()
}
```

can be more clear than:
```rust
fn... {
  let f = if some_condition { futures::future::err("Broken") } else { futures::future::ok(()) };
  f.and_then(|()| some_complex_future_expression())
}
```
which is the dance you need to go through to make there be a single return type.

But also, some usages will just be legacy from before `impl Future` was a thing. So if you're writing new code, aim for `impl Future`, but if you're reading old code, there may not be a good reason it used `BoxFuture` :)

Reply:
The other reason we chose not to rewrite a bunch of code to use `impl Future` is that `async fn` is going to replace almost all usage of `impl Future`. See #6076 for some more discussion.